### DISCUSSION TITLE
Synchronizing 2+ cameras

### DISCUSSION THREAD

**Discussion by athp18** [syntalos/syntalos#12] [Q&A]
Hi!

Thank you again for this very excellent tool, as well as for the module documentation. We were able to develop our own module for recording depth videos with an Orbbec Femto Bolt using the Orbbec SDK, and it works quite well. The current sensor is able to record both infrared and depth videos, and when recording from both simultaneously (without any other module), we don't see any problem and the sensors record stably at our preferred fps. One error that I've noticed however when synchronizing with the UCLA miniscope is that if the Femto Bolt is recording in _both_ infrared and depth, the synchronization tends to get thrown off after a while -- the display and stream FPS for our videos start to fluctuate and decrease, and in the terminal I can see that Syntalos is changing the connection heat level quite regularly. When recording from either IR or depth and synchronizing with the miniscope, however, we don't see any errors, which is great for us :)

I'm curious on if you have any thoughts on this. I suspect it might be related to our system's resources, but not entirely sure. Thank you again for the great, well-documented code! It's been a pleasure to use Syntalos.

**Comment by ximion** [comment]
Hi! This sounds amazing! Congratulations on the module, that's very nice work given how barebones the C++ module development tutorial currently is!

Feel free to contribute the module to the source if you like, so it can be ensured that it remains working!

Just to make sure I understand, you Orbbec module is generating frames that are then saved as video by the `videorecorder` module? Or are you saving them directly? I assume the depth recording might have a format that OpenCV matrices can't represent.

> I can see that Syntalos is changing the connection heat level quite regularly.

This fact should never influence timestamps, because at the time when data is queued, it should already have its timestamp. Syntalos will also heavily deprioritize any data-displaying module if it detects that the module can not keep up, to ensure more resources are available for data-producing modules.

To know if anything can be optimized here, it would be good to know your connection layout :-)

> I'm curious on if you have any thoughts on this. I suspect it might be related to our system's resources, but not entirely sure. Thank you again for the great, well-documented code! It's been a pleasure to use Syntalos.

Thank you! :-)

Time synchronization is actually something that Syntalos should excel at. All timestamps should be aligned to one clock, which in our case is the one of the computer. The data-producing modules have to cooperate though in order to make time synchronization work, and sync information is saved in `tsync` files which you have to read in your later offline analysis (it's a simple binary format documented [here](https://syntalos.readthedocs.io/latest/tsync-format.html), but if you are using Python, you can simply read it using the `edlio` module. Or transform the files into CSV textfiles).

One thing that you should contemplate for your module if you haven't done so yet, is to give it the `REALTIME` feature:
```cpp
ModuleFeatures features() const override
{
    return ModuleFeature::REALTIME | ModuleFeature::SHOW_SETTINGS;
}
```

This will bump the priority of the module's thread when using `void runThread(OptionalWaitCondition *waitCondition) override` and slightly prioritize the module in case it is in an event-looped driver scheme. You can check if you are using a dedicated module thread by seeing if the module's `driver` property is set to return `ModuleDriverKind::THREAD_DEDICATED`.

With the basic stuff out of the way, to align timestamps, you can request a Synchronizer object from Syntalos. Since your device is giving timestamps (and it doesn't just count frames), you will want a `ClockSynchronizer`:
```cpp
// set up clock synchronizer
const auto clockSync = initClockSynchronizer(m_camera->getFPS());
clockSync->setStrategies(TimeSyncStrategy::SHIFT_TIMESTAMPS_FWD | TimeSyncStrategy::SHIFT_TIMESTAMPS_BWD);

// start the synchronizer
if (!clockSync->start()) {
    raiseError(QStringLiteral("Unable to set up clock synchronizer!"));
    return;
}
```

Strategies set how you allow Syntalos to correct the timstamp of your secondary lock (= the clock of your camera device).
For initialization, you need to give the synchronizer the (roughly) expected framerate of your device. It should be able to cope with frame drops just fine though.

How you use the synchronizer now strongly depends on your device: Ideally, the device will give you two timestamps, one when a frame was acquired according to the device's clock, and one when the computer has handled it.
Some devices are dumb and need to be polled by the client, for which Syntalos provides the `MTIMER_FUNC_TIMESTAMP` macro to get the master clock timestamp when the polling function was executed. See https://github.com/syntalos/syntalos/blob/master/modules/camera-tis/tiscameramodule.cpp#L213 for one particularly annoying example (GStreamer peculiarities...).

Once you have both timestamps, you can feed them into the synchronizer:
```cpp 
clockSync->processTimestamp(masterTimeUsec, deviceTimeUsec);
```
This will edit the master-time variable to reflect the time your frame was actually taken. To do that, some statistical analysis is performed to be robust to outliers, but fix up divergence between the computer clock and the clock of your external device.

For a new module using this, it is a good idea to test synchronization by just emitting a stimulus of defined length and defined interval (e.g. blinking a LED, or recording a TTL pulse) to make sure synchronization works well. Once it works, it will work though and you can rely on the synchronization.

Once you are done with it, don't forget to dispose of the synchronizer instance using `safeStopSynchronizer(clockSync);`.

You can check the status of all active synchronizers Syntalos knows about by clicking on *Diagnostics → Timings*. There you can see which offsets are currently corrected for, and what tolerances for time values Syntalos assumes.

It may look like this:
![grafik](https://github.com/user-attachments/assets/3606bf00-0622-4893-890b-0547fa2d7de5)
(but hopefully with less bad offsets, this is a particularly long recording where we messed with the temperature of some devices to make their clocks a bit worse).

There's probably some more tips and tricks, especially for start-time synchronization, but this should get you started! (And further questions are easier to answer if I know how your module looks like and how the Orbbec APIs are designed).

**Reply by athp18** [reply]
Hi! Thanks for the great reply. To answer some of your questions,

> Just to make sure I understand, you Orbbec module is generating frames that are then saved as video by the videorecorder module? Or are you saving them directly? I assume the depth recording might have a format that OpenCV matrices can't represent.

I'm saving the videos using the video recorder module -- the depth videos can be saved properly using OpenCV matrices and I'm able to view them later on later!

> How you use the synchronizer now strongly depends on your device: Ideally, the device will give you two timestamps, one when a frame was acquired according to the device's clock, and one when the computer has handled it.Some devices are dumb and need to be polled by the client, for which Syntalos provides the MTIMER_FUNC_TIMESTAMP macro to get the master clock timestamp when the polling function was executed. See https://github.com/syntalos/syntalos/blob/master/modules/camera-tis/tiscameramodule.cpp#L213 for one particularly annoying example (GStreamer peculiarities...).

I have tried this -- and the m_syTimer->timeSinceStartUsec() function -- and this problem still persists :(

> To know if anything can be optimized here, it would be good to know your connection layout :-)

I'm not using Syntalos input ports (maybe part of the problem?) -- I'm directly interfacing with the Orbbec SDK and my module has two output ports (depth and IR). I'm using the synchronizing functions provided by Syntalos to synchronize the timestamps. I based my module a lot on the example and on the Miniscope one -- you can see the full code here: https://github.com/llorente-lab/syntalos/blob/master/modules/orbbec-module/orbbecmodule.cpp

One thing you might observe is the time sync strategy is set only to backward -- this is because I observed that allowing forward OR backwards was leading to a bunch of console messages along the lines of "timestamp moved backwards when calculating adjusted new time (mitigated by reusing previous time)". 

> This fact should never influence timestamps, because at the time when data is queued, it should already have its timestamp. Syntalos will also heavily deprioritize any data-displaying module if it detects that the module can not keep up, to ensure more resources are available for data-producing modules.

I definitely understand this. IMO the bigger worry for me is that when I open the Canvas module, I'm seeing pretty significant fluctuations in the Stream FPS (as in it drops sometimes to <10 FPS). Additionally, when I open up the timing information, I see something like this:
![Screenshot from 2024-08-05 07-58-16](https://github.com/user-attachments/assets/9444576d-829f-41db-9db1-879a87c22d9d)
The pretty sizable offset in the Depth sensor module worries me a little bit. 
One more thing is that occasionally I will get a flag saying: "a module is overwhelmed with its input".

Also, is it possible that the problem is that this module is accepting *two* video streams (one infrared, one depth) which is just consuming more energy than Syntalos allots for it? If so, is there a way around it? (Part of the problem, IMO, also might be the fact that the Orbbec SDK does take some to initialize and configure the sensors)

I'm not entirely sure if this is expected behavior or if I'm just worrying too much :) But thank you a lot for this great system and all of the resources, as well as for your help, it's helping our lab a lot!

**Reply by athp18** [reply]
As a quick follow-up -- and this is really more related to the edlio package than Syntalos -- if I were to look at timestamps from two simultaneous recordings (say a miniscope and a camera) and load their timestamps, the frame numbers are the same and I could theoretically just quantify the difference through basic subtraction?

**Reply by athp18** [reply]
Any suggestions would be very helpful :)

**Reply by ximion** [reply]
I'll have to have a deeper look at your code and currently have to keep so many plates spinning that I didn't get to it yet. But I should have some time to look into it tomorrow :-)

"a module is overwhelmed with its input" is not necessarily bad (it just means an excessive amount of stuff is buffered and Syntalos will start to monitor the available memory a lot more closely, to emergency-stop the run if needed), but if that message never goes away it means that some module takes too long to process the incoming volume of data.

Usually this is some data-saving module like the *Video Recorder*, choosing a weaker/faster encoder, switching to GPU encoding or upgrading the CPU to one that supports AVX512 may help. The fastest solution if enough storage space is available on disk is to enable to encode the video after a run has completed (then this does not have to happen in realtime and can run in the background).

Shifting times forwards and backwards should work - previously, a BWD shift would mean that the time may run backwards, but nowadays it just means that in extreme cases you may get frames with the same timestamps in case of incredibly unreliable clocks. The behavior you see is odd, I'll have to review your code.

Start time offsets should be mostly constant and - if everything is working properly - sub-sampling. However, it makes sense to measure this a few times (*Tools* -> *Interval Run* is a helpful tool for that). However, just subtracting the times should be a fairly good approximation. I'd still measure it though :-) (hopefully we can soon publish code for that, so it's even easier).

**Reply by athp18** [reply]
Yeah! Of course, take your time. These are very helpful suggestions, I plan to take a look at them and incorporate them. Thank you a lot for all the help so far!

**Reply by athp18** [reply]
I'm curious on if you think the way I'm handling frame processing might be influencing the offsets -- right now, I'm doing OpenCV matrix operations and applying a colormap, _then_ converting it to a Frame struct and pushing it to output -- I can see the miniscope module e.g. sends the raw frame to the Canvas module -- maybe the overhead from processing is causing it?

**Reply by ximion** [reply]
Sorry, I still didn't get to looking at the code... The processing overhead you describe should not cause the "A module is overwhelmed" warning. This warning happens exclusively if a module that is accepting data (usually some kind of sink, like *Canvas* or *Video Recorder* can't keep up with the data-producer. Since they are independent, this can happen.

If the message goes away, I wouldn't worry about it at all :-)

**Reply by ximion** [reply]
Hi, and sorry for the long delay!
Here are a few observations:

* The module-id is [orbbec-cam](https://github.com/llorente-lab/syntalos/blob/master/modules/orbbec-module/orbbecmodule.cpp#L485), but the directory name for the module is `orbbec-module`. Ideally those names should match, I don't think anything will fail right now, but Syntalos generally makes that assumption. Cameras are also prefixed `camera-` to make sorting them easier, so `camera-orbbec` would be a good name, but that's of course optional and not a fixed naming convention.
* https://github.com/llorente-lab/syntalos/blob/master/modules/orbbec-module/orbbecmodule.cpp#L161 - Generally anything can be add as stream metadata, but for camera-type devices you need "framerate", "size" and ideally "has_color". Check https://syntalos.org/docs/modules/camera-arv/ for an example. (And check https://github.com/syntalos/syntalos/blob/master/modules/camera-arv/araviscameramodule.cpp#L118 for how to set that stuff)
* Syntalos can handle 16-bit images, but the problem is many video codecs have issues with it. So currently, while 16-bit images can be passed around within Syntalos, be careful with the video recorder storing the data, because it might be lowered to 8-bit. The Canvas should also be able to handle 16-bit natively (maybe update Syntalos if that's not the case, there were some changes pre-2.0 on that front)
* https://github.com/llorente-lab/syntalos/blob/master/modules/orbbec-module/orbbecmodule.cpp#L227 - You don't need to set the subject ID and especially not the start time. That is stored in the main TOML `attributes.toml` file. And the time you fetch in the *prepare* step will be close to the actual start time, but it won't be the launch time (it will be a bit earlier). Of course, it's also fine to have the JSON file in this way if you data processing pipeline needs these keys :-) (you could fetch the start time from Syntalos later though, I guess, and store it the the file - use `m_syTimer->timeSinceStartMsec()` and subtract that from the current time)
* https://github.com/llorente-lab/syntalos/blob/master/modules/orbbec-module/orbbecmodule.cpp#L236 - You most certainly want to shift FWD and BWD here :-)
* Please move the `initClockSynchronizer` block into `runThread`, so the clock synchronizer only exists in the thread, and no where else. The synchronizer is generally threadsafe, but by having it in the thread you can ensure it is destroyed properly and nothing else uses it. As soon as a run has finished, the synchronizer is destroyed by the Syntalos engine, so you really don't want to have it around somewhere. So: 
```cpp
void runThread(OptionalWaitCondition *waitCondition) final
{
    // set up clock synchronizer
    const auto clockSync = initClockSynchronizer(m_camera->getFPS());
    clockSync->setStrategies(TimeSyncStrategy::SHIFT_TIMESTAMPS_FWD | TimeSyncStrategy::SHIFT_TIMESTAMPS_BWD);

    // start the synchronizer
    if (!clockSync->start()) {
        raiseError(QStringLiteral("Unable to set up clock synchronizer!"));
        return;
    }

    while (m_running) {
        ...
    }

    safeStopSynchronizer(clockSync);
}
```
* If you use functions, just pass the synchronizer around by-reference, (`std::unique_ptr<SecondaryClockSynchronizer> &clockSync`), that should work.
* https://github.com/llorente-lab/syntalos/blob/master/modules/orbbec-module/orbbecmodule.cpp#L413 - you use the synchronizer correctly here :-)
* BUT, you want this function https://github.com/llorente-lab/syntalos/blob/master/modules/orbbec-module/orbbecmodule.cpp#L290 to probably run immediately after a frame appeared, so most likely here https://github.com/llorente-lab/syntalos/blob/master/modules/orbbec-module/orbbecmodule.cpp#L273
* Can you please double-triple-check that the timestamp you get from Orbbec is *indeed* in microseconds? If the two timestamps are in different units, synchronization will produce nonsense, but it will not detect this issue and throw an error. Time unit conversion errors are really annoying and unfortunately frequent.
* Can you check if the frames from Orbbec maybe have *two* timestamps? Some cameras very nicely provide a system timestamp and a device timestamp. If you have both, synchronization becomes a lot more precise as we have a bit more information and Syntalos needs to guess less.
* You could replace this line https://github.com/llorente-lab/syntalos/blob/master/modules/orbbec-module/orbbecmodule.cpp#L272 with this code:
```cpp
FrameSetType frameSet;
auto depthFrameRecvTime = FUNC_DONE_TIMESTAMP(
    m_syTimer->startTime(),
    frameSet = m_pipeline->waitForFrames(200) // wait 200 ms for frames
);
```

What's the current status? (since it has been a while ^^)
